File: 10001st_prime/main.go
package main

import (
	"fmt"
	"math"
)

func main() {
	prime := 2
	for i := 1; i < 10_001; i++ {
		prime = nextPrime(prime)
	}

	fmt.Println(prime)
}

func nextPrime(x int) int {
	for x += 1; ; x++ {
		if isPrime(x) {
			return x
		}
	}

}

func isPrime(x int) bool {
	limit := int(math.Sqrt(float64(x)))
	for i := 2; i <= limit; i++ {
		if x%i == 0 {
			return false
		}
	}
	return true
}

File: 10001st_prime/main.hs
main = print $ findNthPrime 10001

findNthPrime :: Integer -> Integer
findNthPrime n = findNthPrime' 2 (n - 1)

findNthPrime' :: Integer -> Integer -> Integer
findNthPrime' prime 0 = prime
findNthPrime' prime n = findNthPrime' (nextPrime prime) (n - 1)

nextPrime :: Integer -> Integer
nextPrime x
    | isPrime incr = incr
    | otherwise = nextPrime incr
    where incr = succ x

isPrime :: Integer -> Bool
isPrime x = not $ True `elem` results
    where results = map (isDivisible x) [2..isqrt x]

isDivisible :: Integer -> Integer -> Bool
isDivisible x divisor = x `mod` divisor == 0

isqrt :: Integer -> Integer
isqrt = floor . sqrt . fromIntegral